### ISSUE TITLE
Fails to build with node.js 0.12.0

### ISSUE THREAD

**Issue by unbornchikken** [node-ffi/node-ffi#187]
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/gabor/.nvm/versions/node/v0.12.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:267:23)
gyp ERR! stack     at ChildProcess.emit (events.js:110:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1067:12)
gyp ERR! System Linux 3.13.0-37-generic
gyp ERR! command "node" "/home/gabor/.nvm/versions/node/v0.12.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/gabor/git/NOOOCL/node_modules/ffi
gyp ERR! node -v v0.12.0
gyp ERR! node-gyp -v v1.0.2
gyp ERR! not ok 
npm ERR! Linux 3.13.0-37-generic
npm ERR! argv "/home/gabor/.nvm/versions/node/v0.12.0/bin/node" "/home/gabor/.nvm/versions/node/v0.12.0/bin/npm" "update" "--save"
npm ERR! node v0.12.0
npm ERR! npm  v2.5.1
npm ERR! code ELIFECYCLE

**Comment by LinusU** [comment]
This is most likely due to the newer v8 engine being used.

`error: unknown type name 'Arguments'; did you mean 'v8::internal::Arguments'?`
`error: calling a protected constructor of class 'v8::HandleScope'`
`error: unexpected namespace name 'Buffer': expected expression`
`error: no member named 'Close' in 'v8::HandleScope'`
`error: too few arguments to function call, single argument 'isolate' was not specified`
`error: no member named 'NewSymbol' in 'v8::String'`
`error: no matching function for call to 'NODE_SET_METHOD'`

@TooTallNate how would you feel about using [nan](https://github.com/rvagg/nan) to fix this?

**Comment by VitalyTomilov** [comment]
+1

Same here, can no longer compile FFI on Windows 8.1 64-bit after installing Node JS 0.12.0.

Is there any workaround for this?

**Comment by unbornchikken** [comment]
Nope. Above v8 invocations should be converted to NAN ones to make ffi working with Node 0.12 and io.js.

**Comment by TooTallNate** [comment]
I'm down with using nan for this. There's at least 2 PRs already opened for upgrading to nan, and I'm ashamed to say that I've been having trouble having the time to merge them (plus failing tests don't make it any easier). I'm hoping to get this resolved by the end of the week!

**Comment by unbornchikken** [comment]
Thanks for your effort and sorry for the inpatience. I hope that using NAN will resolve the io.js build issue also.

**Comment by TooTallNate** [comment]
Adding io.js support is a priority for me as well :)

**Comment by kjunichi** [comment]
Hi.
I patched to work with node v0.12.0 and io.js v1.1.0.


https://github.com/kjunichi/node-ffi/commit/5ab6bcb8d02da1ab293ef4e1b7abbe8869c61150

this is base on @avatar5d 's work.

Thank you.

**Comment by GCheung55** [comment]
+1

**Comment by CKarper** [comment]
I'm also waiting on this. kjunichi's patch did not build for me...

**Comment by kjunichi** [comment]
Hi,@CKarper.

Thast's too bad.

What OS did you use.
What said your compiler.
Let's me know that.

Thank you.

**Comment by CKarper** [comment]
Windows 7, VS2013 Update 4.   I don't remember the error.  I'll pull
it again and give it a shot. :-)

**Comment by CKarper** [comment]
Some linker errors...   Looks like duplicate function declarations..   Not sure, didn't delve.  Here's the npm output:   http://pastebin.com/5xSw9cmH

**Comment by kjunichi** [comment]
@CKarper thnak you for uploading your log.

I'm afraid to say I have no Windows 7 machine to development.

**Comment by CKarper** [comment]
Now it's looking for pthread.h which isn't present...

http://pastebin.com/SbafsXAT

**Comment by CKarper** [comment]
Looks like you've removed the:
`            , 'deps/pthreads-win32/pthread.gyp:pthread'
from your bindings.gyp?

**Comment by CKarper** [comment]
Yep, if I add that back in, it works.  I sent you a PR.

**Comment by kjunichi** [comment]
I saw this commit.
https://github.com/ghostoy/node-ffi/commit/0324a78c2249cdc2c48d55652b1eac26bde770e0

I see your logs.I merge your req.

Thank you for your pull request.

**Comment by unbornchikken** [comment]
Could you make a PR from this to the main repo? It makes Nate's merging job easier.

**Comment by kjunichi** [comment]
Your comment supports me to try to make PR.
I'm done.

**Comment by unbornchikken** [comment]
I've tested your PR, but there is an issue with it: callbacks doesn't get called.

**Comment by kjunichi** [comment]
Thanks for testing.

Did you say about "npm test"?
I hope you to solve the problems.
I have no idea just now.

I've wirtten some codes that can call from C language to JavaScript successfully.

**Comment by unbornchikken** [comment]
I'm into this but sadly I don't have too mutch v8 native skills yet. CallbackInfo::DispatchToV8 function is not get invoked ... somehow. That's the problem.

**Comment by unbornchikken** [comment]
The issue is the following: in CallbackInfo::Invoke ThreadedCallbackInvocation gets enqueued, a message gets send to the main thread with uv_async_send, but inv->WaitForExecution() never ends because of a deadlock, it looks like there is noone who calls CallbackInfo::WatcherCallback.

**Comment by unbornchikken** [comment]
Oh, I realized that I've tested this with io.js 1.2 on Windows 8 x64 instead of node 0.12. I'll to test it with node later.

**Comment by rosen-vladimirov** [comment]
I've tested it with Nodejs 0.12 and same is happening as described from @unbornchikken with io.js 1.2. I do not mean npm test is not working - our code, which is based on ffi calls is not working :(

**Comment by unbornchikken** [comment]
Yes, I forgot to mention that my ffi based tests got deadlocked. I'm working on an OpenCL bindings module: https://github.com/unbornchikken/NOOOCL, and my tests cause ffi deadlocks. You can try them, they are in the beta-dev branch.

 I was able to test 0.12 and io,js 1.2 on Linux with @kjunichi 's PR https://github.com/node-ffi/node-ffi/pull/189. There is no deadlocks in my module, but @TooTallNate 's tests still fails by not throwing expected exceptions errors, like in the Travis CI logs. So I think this should be a libuv upgrade related issue.

**Comment by CKarper** [comment]
The fix that I reversed by adding pthread back in was a thread deadlocking
fix.  It's likely that this is what they were seeing.  Maybe it works
without pthread being included, but I was getting link errors, so there was
some kind of hole on their part.   I would check that again, and maybe this
is an already fixed issue.

**Comment by unbornchikken** [comment]
pthread is used for synchronization in this module, so cannot be removed so easily. I think it could be replaced by libuv, it already supports necessary synchronization constructs. But I think this issue is caused by libuv not

**Comment by rosen-vladimirov** [comment]
Hey guys, any progress on this? I'm ready to help with testing (my V8 knowledge is poor, so I don't think I can help you with code).

**Comment by kjunichi** [comment]
Hi, rosen-vladimirov

Progress is here.

- https://github.com/node-ffi/node-ffi/pull/189

**Comment by unbornchikken** [comment]
It's cool, man. I'll test this on Linux and Windows today with my own library's unit tests and reply.

**Comment by unbornchikken** [comment]
Ok, I've tested it. It finally deadlock free on windows. May I ask why have you disabled failing unit tests?

**Comment by kjunichi** [comment]
It seems that I have no error to merge this patch.



2015/02/20 22:17、Gábor Mező <notifications@github.com> のメッセージ:

**Comment by unbornchikken** [comment]
I mean there are a bounch of disabled ffi unit tests with it.skip. If I reenable them those still fails.

**Comment by VitalyTomilov** [comment]
God bless your efforts, guys. I probably would speak for many others, sitting here on couple large projects, being unable to upgrade to Node 0.12.0 or IO 1.2 just because of FFI.

Can't wait for the fix code merge and the new release.

**Comment by unbornchikken** [comment]
You can go with https://github.com/node-ffi/node-ffi/pull/189 from @kjunichi . I don't have any issues with my own unit tests in my above mentioned OpenCL wrapper library by using it.

Just put this into dependencies in package.json:

    "ffi": "git+https://github.com/kjunichi/node-ffi.git#bad796888ad564bbc6a15b73579223b8f118605e"

**Comment by kjunichi** [comment]
I mean that.Some tests fail because of libuv, nothing with libffi.


i thought that tests have some errors, this patch never to be merged.
So I skip the test.

Now it seems that npm on win32 platforms has some problems.

There is a last one step.

2015/02/20 23:19、Gábor Mező <notifications@github.com> のメッセージ:

**Comment by VitalyTomilov** [comment]
Closed both this one and the referenced issue. I'm not seeing any commits fixing the issue. How come it is closed now?

**Comment by rosen-vladimirov** [comment]
Hi @kjunichi , 
we are testing your code and for the moment it seems to work as expected. Thanks for the great work and your efforts. I'll write back when we finish our testing.

**Comment by VitalyTomilov** [comment]
When should we expect a new release with the fix?

**Comment by rosen-vladimirov** [comment]
Hi again @kjunichi
Thank you very much for your efforts. From our tests we can say that your version is working fine with node 0.12 and node 0.10.36
We'll use it on our side. Great work (special thanks to @avatar5d  as well)

**Comment by gogins** [comment]
Could you please integrate this with the release? This is something I urgently need.

**Comment by vitaly-t** [comment]
+1, please somebody, release the fix, too many people are waiting for it.

**Comment by daemon-7** [comment]
this is my err info.

npm install -g ffi --msvs_version=2012 --save-dev

gyp ERR! build error
gyp ERR! stack Error: `C:\Windows\Microsoft.NET\Framework\v4.0.30319\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (F:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:267:23)
gyp ERR! stack     at ChildProcess.emit (events.js:110:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1067:12)
gyp ERR! System Windows_NT 6.1.7601
gyp ERR! command "node" "F:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\admin\AppData\Roaming\npm\node_modules\ffi
gyp ERR! node -v v0.12.0
gyp ERR! node-gyp -v v1.0.2
gyp ERR! not ok
npm ERR! Windows_NT 6.1.7601
npm ERR! argv "F:\\Program Files\\nodejs\\\\node.exe" "F:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "-g" "ffi" "--msvs_version=2012" "--save-dev"
npm ERR! node v0.12.0
npm ERR! npm  v2.5.1
npm ERR! code ELIFECYCLE

npm ERR! ffi@1.2.7 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the ffi@1.2.7 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the ffi package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls ffi
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     D:\nw.js\nw_native_dll_sample-master\app\npm-debug.log

**Comment by opadron** [comment]
+1, this issue is blocking a number of our in-house projects

[npm-debug log](http://pastebin.com/raw.php?i=rdfCyYnb)

**Comment by Cerealkillerway** [comment]
+1

**Comment by unbornchikken** [comment]
I think @kjunichi's PR is not ready to get merged into the master, because there are many failing test cases that @TooTallNate put there for a good reason. What @kjunichi did is that he simply disabled failing test so there is no Travic CI complaints in his PR. But we all know this is not the correct way to fix failing test cases, right? ;)

Let me help you with this. I've forked @kjunichi's version, and restored original test cases. There was a serious bug in @TooTallNate's original test code that caused mocha test flow crash in the middle of the testing process. That was the cause of the half of the failing tests cases. I'm fixed it in https://github.com/unbornchikken/node-ffi/commit/8b37f7cbf1566855c2a45526fd4e2c29c6dc32bb

So there is only one issue left that prevents **all of** @TooTallNate's tests passing: exceptions doesn't get propagated back to JS side from the ffi side. There are 7 failed tests but the cause of failing is always this.

I'll try to look into this.

**Comment by kjunichi** [comment]
+1
Nice work!

**Comment by LinusU** [comment]
@unbornchikken for a greater chans of getting your work merged I would suggest to not change the coding style. Also, shouldn't the listeners be added back inside a finally block? (or just outside of the try catch)

```javascript
var err;

try {
  assert(/ffi/.test(e.message))
} catch (_err) {
  err = _err
}

listeners.forEach(function (fn) {
  process.on('uncaughtException', fn)
})

done(err)
```

The `done` call shouldn't be in the `catch` since that will run before `finally`. Or, we could just send the error to mocha thru `uncaughtException`:

```javascript
listeners.forEach(function (fn) {
  process.on('uncaughtException', fn)
})

assert(/ffi/.test(e.message))
done()
```

**Comment by unbornchikken** [comment]
I'm trying to configure my WebStorm to reflect the coding style that @TooTallNate used in the first place, please check out my following commit, so if I can manage to create a PR from this, there shouldn't be whitespace diffs for sure.

Nice catch is the above complaint, will be fixed in the next commit.

By the way my current progress is: 4 failing tests remains, was 7.

**Comment by unbornchikken** [comment]
Ok, I did my last commit today, only one failing test remains. I can fix that too, as soon as I have a few hours. Then we will have an ffi, that has all unit test from @TooTallNate passed. Till Nate has time to review and merge, I create a temporary module from it on the npm, so you guys can move towards io.js or 0.12 easily.

**Comment by vitaly-t** [comment]
@unbornchikken, great work, we can't wait to see this making a new release ;)

**Comment by unbornchikken** [comment]
It's done: https://www.npmjs.com/package/ffi-io

All tests passed, tested on: Linux, Windows with Node 0.10+, io.js 1.5.1. I've opened the Issue tracker on my repo: https://github.com/unbornchikken/node-ffi, and try to support this version in my (sparse) spare time till @TooTallNate able to continue his work on the original ffi package there.

Making of the PR is in progress.

**Comment by kjunichi** [comment]
+1

What a great work!

FYI
I test it on 
```
sw_vers 
ProductName:    Mac OS X
ProductVersion: 10.10.2
BuildVersion:   14C1510
```



```
$ npm test

> ffi-io@1.0.2 test /Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io
> node-gyp rebuild --directory test && mocha -gc --reporter spec

child_process: customFds option is deprecated, use stdio instead.
  CXX(target) Release/obj.target/ffi_tests/ffi_tests.o
  SOLINK_MODULE(target) Release/ffi_tests.node
  SOLINK_MODULE(target) Release/ffi_tests.node: Finished


  Callback
    ✓ should create a C function pointer from a JS function 
    ✓ should be invokable by an ffi'd ForeignFunction 
    ✓ should work with a "void" return type 
    ✓ should not call "set()" of a pointer type 
    ✓ should throw an Error when invoked through a ForeignFunction and throws 
    ✓ should throw an Error with a meaningful message when a type's "set()" throws 
    ✓ should throw an Error when invoked after the callback gets garbage collected 
    async
      ✓ should be invokable asynchronously by an ffi'd ForeignFunction 
      ✓ multiple callback invocations from uv thread pool should be properly synchronized (1769ms)
      ✓ should work being invoked multiple times (94ms)
      ✓ should throw an Error when invoked after the callback gets garbage collected 

  ffi_cif
    ✓ should return a Buffer representing the `ffi_cif` instance 
    ✓ should throw an Error when given an invalid "type" 
    ✓ should throw an Error when given an invalid ABI argument 

  DynamicLibrary
    ✓ should be a function 
    ✓ should return a "DynamicLibrary" instance when invoked (42ms)
    get()
      ✓ should return a "pointer" Buffer to a symbol 
      ✓ should set the "name" property to the name of the symbol 

  errno()
    ✓ should be a function 
    ✓ should set the errno with out-of-range "strtoul" value 

  ForeignFunction
    ✓ should call the static "abs" bindings 
    ✓ should throw an Error with a meaningful message when type's `set()` throws 
    ✓ should call the static "atoi" bindings 
    ✓ should call the static "double_box" bindings 
    ✓ should call the static "double_box_ptr" bindings 
    ✓ should call the static "area_box" bindings 
    ✓ should call the static "area_box_ptr" bindings 
    ✓ should call the static "create_box" bindings 
    ✓ should call the static "add_boxes" bindings 
    ✓ should call the static "int_array" bindings 
    ✓ should call the static "array_in_struct" bindings 
    ✓ should not call the "ref()" function of its arguments 
    async
      ✓ should call the static "abs" bindings asynchronously 
      ✓ should invoke the callback with an Error with a meaningful message when type's `set()` throws 

  Function "type"
    ✓ should be a function 
    ✓ should return a "type" object when invoked with a return type and array of arguments types 
    ✓ should be accepted as a return "type" to a ForeignFunction 
    ✓ should be accepted as an argument "type" to a ForeignFunction 
    ✓ should work as expected using the "callback_func" static bindings 

  Library
    ✓ should be a function 
    ✓ should work with the `new` operator 
    ✓ should accept `null` as a first argument 
    ✓ should accept a lib name as the first argument 
    ✓ should accept a lib name with file extension 
    ✓ should throw when an invalid function name is used 
    ✓ should work with "strcpy" and a 128 length string 
    ✓ should work with "strcpy" and a 2k length string (42ms)
    ✓ should work with "gettimeofday" and a "timeval" Struct pointer 
    async
      ✓ should call a function asynchronously 

  @try / @catch
    ✓ should proxy @try/@catch to JavaScript via try/catch/throw 
    1) should throw a Buffer instance when an exception happens

  types
    `ffi_type` to ref type matchups
      ✓ should match a valid `ffi_type` for "void" 
      ✓ should match a valid `ffi_type` for "int8" 
      ✓ should match a valid `ffi_type` for "uint8" 
      ✓ should match a valid `ffi_type` for "int16" 
      ✓ should match a valid `ffi_type` for "uint16" 
      ✓ should match a valid `ffi_type` for "int32" 
      ✓ should match a valid `ffi_type` for "uint32" 
      ✓ should match a valid `ffi_type` for "int64" 
      ✓ should match a valid `ffi_type` for "uint64" 
      ✓ should match a valid `ffi_type` for "float" 
      ✓ should match a valid `ffi_type` for "double" 
      ✓ should match a valid `ffi_type` for "Object" 
      ✓ should match a valid `ffi_type` for "CString" 
      ✓ should match a valid `ffi_type` for "bool" 
      ✓ should match a valid `ffi_type` for "byte" 
      ✓ should match a valid `ffi_type` for "char" 
      ✓ should match a valid `ffi_type` for "uchar" 
      ✓ should match a valid `ffi_type` for "short" 
      ✓ should match a valid `ffi_type` for "ushort" 
      ✓ should match a valid `ffi_type` for "int" 
      ✓ should match a valid `ffi_type` for "uint" 
      ✓ should match a valid `ffi_type` for "long" 
      ✓ should match a valid `ffi_type` for "ulong" 
      ✓ should match a valid `ffi_type` for "longlong" 
      ✓ should match a valid `ffi_type` for "ulonglong" 
      ✓ should match a valid `ffi_type` for "size_t" 
      ✓ should match a valid `ffi_type` for "ref" type without a cached value 
      ✓ should match a valid `ffi_type` for `CString` without a cached value 
      ✓ should match a valid `ffi_type` for `ulong` without a cached value 

  variadic arguments
    ✓ should work with vararg C functions 
    ✓ should return the same Function instance when the same arguments are used 


  81 passing (3s)
  1 failing

  1) @try / @catch should throw a Buffer instance when an exception happens:
     AssertionError: false == true
      at Context.<anonymous> (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/test/objc.js:47:9)
      at callFn (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runnable.js:266:21)
      at Test.Runnable.run (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runnable.js:259:7)
      at Runner.runTest (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:387:10)
      at /Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:470:12
      at next (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:312:14)
      at /Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:322:7
      at next (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:257:23)
      at Immediate._onImmediate (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:289:5)
      at processImmediate [as _immediateCallback] (timers.js:358:17)



npm ERR! Test failed.  See above for more details.
```

Thank you!

**Comment by unbornchikken** [comment]
Please put it's issues into my repo's issue list, and I'll look into this. https://github.com/unbornchikken/node-ffi

**Comment by unbornchikken** [comment]
This is an OSX only test case I can see now. Saddly I don't and won't ever have a Mac, I'm a PC, so any help would be greatly appreciated to fixing of this. My gues is that NanThrowError method somehow coerce error values to be Error instances instead of any value.

**Comment by TooTallNate** [comment]
> But this is not critical, even is not a bug, catched errors should be Error instances anyway.

That is incorrect. Buffer instances are what gets thrown because they're pointing to the C++ error object, which your JS code could do _something_ with. For example NodObjC casts them into Objective-C objects and allows you to pass Objc messages to the error instance from there.

**Comment by unbornchikken** [comment]
@kjunichi , we're working on the PR there https://github.com/node-ffi/node-ffi/pull/193. I think my last commit fixes the above OSX issue. I've released the progress of the CR on the npm: https://www.npmjs.com/package/ffi-io , if you have time, please run the test cases of it on OSX. Thanks.

**Comment by kjunichi** [comment]
I'm afraid to say, it's not OK.

```
81 passing (3s)
  1 failing

  1) @try / @catch should throw a Buffer instance when an exception happens:
     AssertionError: false == true
      at Context.<anonymous> (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/test/objc.js:47:9)
      at callFn (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runnable.js:266:21)
      at Test.Runnable.run (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runnable.js:259:7)
      at Runner.runTest (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:387:10)
      at /Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:470:12
      at next (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:312:14)
      at /Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:322:7
      at next (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:257:23)
      at Immediate._onImmediate (/Users/junichi/work/nodejs/ffi-io/node_modules/ffi-io/node_modules/mocha/lib/runner.js:289:5)
      at processImmediate [as _immediateCallback] (timers.js:369:17)



npm ERR! Test failed.  See above for more details.
```

I tested this code add to above tests.
No exception is thrown.

```js
cat test.js
var ref = require('ref');
var ffi = require('ffi-io');
var voidPtr = ref.refType(ref.types.void);

var id = voidPtr;
var SEL = voidPtr;
var Class = voidPtr;


var objcLib = new ffi.Library('libobjc', {
  'objc_msgSend': [ id, [ id, SEL ] ],
  'objc_getClass': [ Class, [ 'string' ] ],
  'sel_registerName': [ SEL, [ 'string' ] ]
});
var NSAutoreleasePool = objcLib.objc_getClass('NSAutoreleasePool');
var sel_new = objcLib.sel_registerName('new');
var pool = objcLib.objc_msgSend(NSAutoreleasePool, sel_new);

var sel_retain = objcLib.sel_registerName('retaian');
try {
  objcLib.objc_msgSend(pool, sel_retain);
  console.log("oops!");
} catch (e) {
  console.dir(e);
}

var sel_retain = objcLib.sel_registerName('retain');
try {
  objcLib.objc_msgSend(pool, sel_retain);
  console.log("oops!");
} catch (e) {
  console.dir(e);
}

var NSString = objcLib.objc_getClass('NSString');
s = objcLib.objc_msgSend (NSString, objcLib.sel_registerName("alaloc"));
```

result is this.

```
node test
oops!
oops!
```

thank you.

**Comment by unbornchikken** [comment]
I'm totally stabbing in the dark here, it was a years ago when I took a look at the screen of a Mac. I'm sure that @TooTallNate will look into this. Till it's good for anyone who are not using ffi for doing Objective C API based stuff.

**Comment by TooTallNate** [comment]
v1.3.0 published to npm

**Comment by TooTallNate** [comment]
@kjunichi strangely enough, I _was_ experiencing that on my iMac, but _not_ on my MacBook Pro... only difference that I can think of is that on the iMac I use `n` for version management, and on the MBP I use `nvm`... but I understand that they download the same precompiled tarball, so it would be strange to me if that was the culprit.

For some reason, it seems like C++ exceptions are disabled for some reason, under some particular circumstance.

**Comment by LinusU** [comment]
That one test is failing for me with `node v0.12.0` installed via Homebrew on MacBook Air 6,2.

```sh
git clone git@github.com:node-ffi/node-ffi.git
cd node-ffi
npm install
npm test
```

```text
  1 failing

  1) @try / @catch should throw a Buffer instance when an exception happens:
     AssertionError: false == true
      at Context.<anonymous> (/private/tmp/node-ffi/test/objc.js:47:9)
      at callFn (/private/tmp/node-ffi/node_modules/mocha/lib/runnable.js:266:21)
      at Test.Runnable.run (/private/tmp/node-ffi/node_modules/mocha/lib/runnable.js:259:7)
      at Runner.runTest (/private/tmp/node-ffi/node_modules/mocha/lib/runner.js:387:10)
      at /private/tmp/node-ffi/node_modules/mocha/lib/runner.js:470:12
      at next (/private/tmp/node-ffi/node_modules/mocha/lib/runner.js:312:14)
      at /private/tmp/node-ffi/node_modules/mocha/lib/runner.js:322:7
      at next (/private/tmp/node-ffi/node_modules/mocha/lib/runner.js:257:23)
      at Immediate._onImmediate (/private/tmp/node-ffi/node_modules/mocha/lib/runner.js:289:5)
      at processImmediate [as _immediateCallback] (timers.js:358:17)
```

**Comment by TooTallNate** [comment]
@kjunichi @LinusU Actually now that I've updated my MBP to the latest Yosemite developer beta build, this test case is failing for me now as well... Not sure what's the deal with that, but it's definitely something related to the operating system disabling C++ exceptions for some reason.

**Comment by kjunichi** [comment]
@TooTallNate thank you for investigation.

I don' know C++ exception well.But I heard libocjc was changed at some moments, so i think this couse by this not throw exception problems. 

BTW 
I also like your other products too.It's wonderful.I like ansi canvas very much.

**Comment by frndxyz** [comment]
getting this error on window 2008 R2 ,node 0.10.25 , VS 2015 ,

  dlfcn.c
  dlfcn.vcxproj -> C:\node_modules\node-ffi\build\Release\\dlfcn.lib
  test
  ..\libffi-build.sh: deps/libffi: No such file or directory
  c:\sh\make.exe: *** No targets.  Stop.
  pthread.c
c:\node_modules\node-ffi\deps\pthreads-win32\pthread.h(320): error C2011: 'timespec': 'struct' type redefinition [C:\no
de_modules\node-ffi\build\deps\pthreads-win32\pthread.vcxproj]
  C:\Program Files (x86)\Windows Kits\10\Include\10.0.10150.0\ucrt\time.h(39): note: see declaration of 'timespec'
gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\MSBuild\14.0\bin\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:2
67:23)
gyp ERR! stack     at ChildProcess.EventEmitter.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:797:12)
gyp ERR! System Windows_NT 6.1.7601
gyp ERR! command "node" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuil
d"
gyp ERR! cwd C:\node_modules\node-ffi
gyp ERR! node -v v0.10.25
gyp ERR! node-gyp -v v0.12.2
gyp ERR! not ok
npm ERR! node-ffi@0.5.7 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the node-ffi@0.5.7 install script.
npm ERR! This is most likely a problem with the node-ffi package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls node-ffi
npm ERR! There is likely additional logging output above.

npm ERR! System Windows_NT 6.1.7601
npm ERR! command "C:\\Program Files\\nodejs\\\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js"
 "install" "node-ffi"
npm ERR! cwd C:\
npm ERR! node -v v0.10.25
npm ERR! npm -v 1.3.24
npm ERR! code ELIFECYCLE
npm ERR!
npm ERR! Additional logging details can be found in:
npm ERR!     C:\npm-debug.log
npm ERR! not ok code 0
